### DISCUSSION TITLE
Firebase

### DISCUSSION THREAD

**Discussion by bryanjtc** [vikejs/vike#601] [Discussion]
Does anyone has a working example with firebase?

**Comment by truuman** [comment]
Yes I have. What would you like to know?

I have a project that's using Realtime Database, Functions, App Check, Authentication and Hosting.

**Reply by bryanjtc** [reply]
How did you set up the firebase function for the server?

**Reply by bryanjtc** [reply]
import functions from "firebase-functions";
import express from "express";
import compression from "compression";
import { renderPage } from "vite-plugin-ssr";
import cors from "cors";
import { dirname } from "path";
import { fileURLToPath } from "url";
const __dirname = dirname(fileURLToPath(import.meta.url));
const root = `${__dirname}`;

const ssrapp = express();

ssrapp.use(compression());
ssrapp.use(cors({ origin: true }));
const sirv = (await import("sirv")).default;
ssrapp.use(sirv(`${root}/dist/client`));
ssrapp.get("*", async (req, res) => {
  const pageContextInit = {
    urlOriginal: req.originalUrl,
  };
  const pageContext = await renderPage(pageContextInit);
  const { httpResponse } = pageContext;
  if (!httpResponse) return next();
  const { body, statusCode, contentType, earlyHints } = httpResponse;
  if (res.writeEarlyHints)
    res.writeEarlyHints({ link: earlyHints.map((e) => e.earlyHintLink) });
  res.status(statusCode).type(contentType).send(body);
});

const app = functions.https.onRequest(ssrapp);
export { app };


This is my function. It keeps throwing an error with vite-plugin-ssr
I copied the dist folder from my vite app to the functions firebase folder.

**Reply by truuman** [reply]
I don't think you are supposed to run functions inside your server. Functions are deployed to Firebase and they are triggered based on the interactions you do with Firebase features, e.g. on a new entry in Firestore or Realtime Database.

In my case, I have an email notification system set up. When someone fills out a form in the vite-plugin-ssr app, a new entry is created in the Realtime Database which triggers a Function that sends an email.

**Reply by bryanjtc** [reply]
The function is to run the server. My app uses server-side-rendering.

**Reply by truuman** [reply]
Wow cool, I didn't know you can actually do that.

Could you post the error you receive from Firebase? Could be something as simple as a node version mismatch. I don't know if you can set the node version for Firebase Functions though.

**Reply by bryanjtc** [reply]
I ended up using the pre-render option and uploading the static files to firebase hosting.

**Reply by truuman** [reply]
Yeah I'm doing the same on my project. I'm interested in running SSR via functions though, I might look into it.
